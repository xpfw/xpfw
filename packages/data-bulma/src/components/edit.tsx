import { dataOptions, IEditHookProps, useEditWithProps } from "@xpfw/data";
import {  getMapToFromProps, iterateSubFields, prependPrefix, SharedField } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite";
import * as React from "react"

const BulmaEdit: React.FunctionComponent<IEditHookProps> = observer((props) => {
  const editProps = useEditWithProps(props)
  const fields: any[] = []
  iterateSubFields(props.schema, (key, schema) => {
    fields.push(<SharedField key={key} schema={schema} prefix={prependPrefix(getMapToFromProps(props), props.prefix)} />)
  })
  const gotErr = editProps.error != null
  const result = editProps.state
  let msg: any
  if (gotErr) {
    msg = (
      <div className="notification is-danger">
        Error please recheck your inputs or connection {JSON.stringify(editProps.error)}
      </div>
    )
  }
  if (result) {
    msg = (
      <div className="notification is-success">
        Successfully saved changes to {get(result, dataOptions.idPath)}
      </div>
    )
  }
  return (
    <div>
      {fields}
      <button onClick={editProps.submitEdit}>Save Changes</button>
      {msg}
    </div>
  )
})

export default BulmaEdit
