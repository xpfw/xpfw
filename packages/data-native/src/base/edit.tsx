import * as React from "react"

import { dataOptions,  useEditWithProps, IEditHookProps } from "@xpfw/data"
import { getMapToFromProps, iterateSubFields, prependPrefix, SharedField } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"
import { View } from "react-native"
import { Button, Text, Card } from "react-native-elements"

const NativeEdit: React.FunctionComponent<IEditHookProps> = observer((props) => {
  const createProps = useEditWithProps(props)
  const fields: any[] = []
  iterateSubFields(props.schema, (key, schema) => {
    fields.push(<SharedField key={key} schema={schema} prefix={prependPrefix(getMapToFromProps(props), props.prefix)} />)
  })
  const gotErr = createProps.error != null
  const result = createProps.state
  let msg: any
  if (gotErr) {
    msg = (
      <Card
        containerStyle={{backgroundColor: "green", margin: 3}}
      >
        <Text style={{color: "white", textAlign: "center"}}>
          Error please recheck your inputs or connection {JSON.stringify(createProps.error)}
        </Text>
      </Card>
    )
  }
  if (result) {
    msg = (
      <Card
        containerStyle={{backgroundColor: "green", margin: 3}}
      >
        <Text style={{color: "white", textAlign: "center"}}>
          Successfully saved changes to {get(result, dataOptions.idPath, "")}
        </Text>
      </Card>
    )
  }
  return (
    <View>
      {fields}
      <View style={{marginTop: 20, marginLeft: 5, marginRight: 5}}>
        <Button onPress={createProps.submitEdit} title="Save" />
      </View>
      {msg}
    </View>
  )
})

export default NativeEdit