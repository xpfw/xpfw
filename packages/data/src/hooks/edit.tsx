import { ExtendedJSONSchema, FormStore, getMapTo, jsonValidator, memo, prependPrefix } from "@xpfw/form"
import { useEffect } from "react"
import DbStore from "../store/db"
import UserStore from "../store/user"
import toJS from "../util/toJS"

const submitEdit = (id: string, schema: ExtendedJSONSchema, mapTo?: string, prefix?: string) => {
  return () => {
    if (mapTo == null) {
      mapTo = getMapTo(schema, mapTo)
    }
    const newValue = toJS(FormStore.getValue(mapTo, prefix))
    jsonValidator.validate(schema, newValue)
    FormStore.setError(mapTo + id, jsonValidator.errors, prefix)
    const err = FormStore.getError(mapTo + id, prefix)
    if (err == null) {
      return DbStore.patch(id, schema, mapTo, prefix)
    }
    return err
  }
}
const u: any = DbStore
const useEdit = (id: string, schema: ExtendedJSONSchema, mapTo?: string, prefix?: string, reset?: boolean) => {
  if (mapTo == null) { mapTo = getMapTo(schema, mapTo) }
  const valuePath = `${prependPrefix(mapTo, prefix)}${id}`
  useEffect(() => {
    if (reset) {
      u.currentlyEditing = ""
      u.updateState[prependPrefix(mapTo, prefix)] = null
      FormStore.setValue(mapTo, {}, prefix)
    }
  }, [reset])
  return {
    error: FormStore.getError(valuePath),
    state: DbStore.getUpdateState(valuePath),
    loading: FormStore.getLoading(valuePath),
    original: DbStore.getEditOriginal(id, schema, mapTo, prefix, false),
    user: UserStore.getUser(),
    submitEdit: memo(() => submitEdit(id, schema, mapTo, prefix),
      ["submitEdit", id, JSON.stringify(schema), mapTo, prefix])
  }
}

export interface IEditHookProps {
  schema: ExtendedJSONSchema
  id: string
  prefix?: string
  mapTo?: string
  reset?: boolean
}

const useEditWithProps = (props: IEditHookProps) => useEdit(props.id, props.schema, props.mapTo, props.prefix, props.reset)

export default useEdit
export {
  useEditWithProps, submitEdit
}
