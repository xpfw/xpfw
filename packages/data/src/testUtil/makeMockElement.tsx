import { cloneDeep, get, isString, set } from "lodash-es"
import * as React from "react"

const idtmp: string[] = []

const makeMockElement = (name: string, stripIdsAt?: string | string[], idPath: string = "_id", stringifier?: () => any) => {
  return class extends React.Component<any, any> {
    public render() {
      const toStringify = stringifier ? stringifier() : this.props
      const stripMe = Array.isArray(stripIdsAt) ? stripIdsAt : [stripIdsAt]
      if (Array.isArray(stripMe)) {
        for (const stripper of stripMe) {
          if (isString(stripper)) {
            let toStrip = get(this.props, stripper)
            if (!Array.isArray(toStrip)) {
              toStrip = [toStrip]
            }
            const newIds = []
            for (const obj of toStrip) {
              const id = isString(obj) ? obj : get(obj, idPath)
              let newId = idtmp.indexOf(id)
              if (newId === -1) {
                newId = idtmp.length + 1
                idtmp.push(id)
              }
              const objClone = cloneDeep(obj)
              if (isString(obj)) {
                newIds.push(newId)
              } else {
                set(objClone, idPath, newId)
                newIds.push(objClone)
              }
            }
            if (newIds.length > 0 && newIds[0] != null) {
              set(toStringify, stripper, newIds)
            }
          }
        }
      }
      return (
        <div className={name}>
          {JSON.stringify(toStringify, undefined, 2)}
        </div>
      )
    }
  }
}

export default makeMockElement
