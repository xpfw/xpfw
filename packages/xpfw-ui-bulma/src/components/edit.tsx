import { FormStore, SharedField } from "@xpfw/form-shared"
import { IFormEditProps, ISharedFormEdit, ListStore, SharedFormEdit } from "@xpfw/ui-shared"
import { getFieldsFromForm,  IField } from "@xpfw/validate"
import { get } from "lodash"
import * as React from "react"

class MiniEdit extends React.Component<IFormEditProps, any> {
  public render() {
    const fields = this.props.fields.map((field: IField) => {
      return <SharedField key={field.mapTo} field={field} prefix={this.props.prefix} user={this.props.user} />
    })
    const gotErr = get(this.props, "error.errors.length", 0)
    const result = get(this.props, "state.result")
    let msg: any
    if (gotErr) {
      msg = (
        <div className="notification is-danger">
          Error please recheck your inputs or connection {JSON.stringify(get(this.props, "error"))}
        </div>
      )
    }
    if (result) {
      msg = (
        <div className="notification is-success">
          Successfully saved changes to {get(result, "_id")}
        </div>
      )
    }
    return (
      <div>
        {fields}
        <button onClick={this.props.submitEdit}>Save Changes</button>
        {msg}
      </div>
    )
  }
}

const BulmaEdit: React.ComponentType<ISharedFormEdit> = SharedFormEdit<{}>(MiniEdit)

export default BulmaEdit
