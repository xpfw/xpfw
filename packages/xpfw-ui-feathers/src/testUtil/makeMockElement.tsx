import * as React from "react"
import { get, set, isString, isArray, cloneDeep } from "lodash";

const idtmp: string[] = []

const makeMockElement = (name: string, stripIdsAt?: string, idPath: string = "_id") => {
  return class extends React.Component<any, any> {
    public render() {
      const toStringify = this.props
      if (isString(stripIdsAt)) {
        let toStrip = get(this.props, stripIdsAt)
        if (!isArray(toStrip)) {
          toStrip = [toStrip]
        }
        const newIds = []
        for (const obj of toStrip) {
          const id = get(obj, idPath)
          let newId = idtmp.indexOf(id)
          if (newId === -1) {
            newId = idtmp.length + 1
            idtmp.push(id)
          }
          const objClone = cloneDeep(obj)
          set(objClone, idPath, newId)
          newIds.push(objClone)
        }
        set(toStringify, stripIdsAt, newIds)
      }
      return (
        <div className={name}>
          {JSON.stringify(toStringify, undefined, 2)}
        </div>
      )
    }
  }
}

export default makeMockElement
