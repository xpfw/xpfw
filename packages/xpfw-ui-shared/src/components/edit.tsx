import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { filterFields, globals, IField, IForm, IFormError } from "@xpfw/validate"
import { get, isNil } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore from "../store/db"
import UserStore from "../store/user"

export interface ISharedFormEdit extends React.Props<any> {
  prefix?: string
  id?: string
  /**
   * Will reset the form values
   */
  resetOnMount?: boolean
  /**
   * Will reset the result of the last creation
   */
  resetState?: boolean
  form: IForm
}

export interface ISharedFormEditState {
  error?: IFormError
  state?: any
  original?: any
  user?: any
  loading: boolean
}

export type SubmitEdit = () => Promise<any>

export interface IFormEditProps extends ISharedFormEdit, ISharedFormEditState {
  submitEdit: SubmitEdit
  /** Prefiltered Fields to display */
  fields: IField[]
}

const submitEdit = (thisRef: {props: ISharedFormEdit}) => {
  return async () => {
    const prefix = get(thisRef, "props.prefix", "")
    await FormStore.validateForm(get(thisRef, "props.form"),
      globals.Method.Update, prefix, get(thisRef, "state.user"))
    const err =
    FormErrorStore.getError(`${prefix && prefix.length > 0 ? prefix + "." : ""}${get(thisRef, "props.form.model")}`)
    let res
    if (!isNil(err) && err.ok) {
      res = await DbStore.patch(get(thisRef, "props.id"),
        get(thisRef, "props.form"), get(thisRef, "props.prefix"))
    }
    return res
  }
}

function SharedFormEdit<T>(Component: React.ComponentType<IFormEditProps & T>):
React.ComponentType<ISharedFormEdit & T> {
  return class extends ComponentBase<ISharedFormEdit & T, ISharedFormEditState> {
    public submitEdit: SubmitEdit
    public constructor(props: ISharedFormEdit & T) {
      super(props)
      this.submitEdit = submitEdit(this)
    }
    public componentWillMount() {
      const prefix: any = get(this.props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(this.props, "id", "")}${get(this.props, "form.model", "undefined")}`
      if (this.props.resetOnMount) {
        FormErrorStore.setError(valuePath, undefined)
      }
      if (this.props.resetState) {
        DbStore.setUpdateState(valuePath, undefined)
      }
      super.componentWillMount()
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitEdit={this.submitEdit}
          fields={filterFields(this.props.form, globals.Method.Update)}
        />
      )
    }
    protected _buildState(props: ISharedFormEdit, initialBuild: boolean): ISharedFormEditState {
      const prefix: any = get(props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(props, "id", "")}${get(props, "form.model", "undefined")}`
      const id: any = get(props, "id", "none")
      const collection = get(props, "form.collection", "none")
      return {
        error: FormErrorStore.getError(valuePath),
        state: DbStore.getUpdateState(valuePath),
        loading: LoadingStore.getLoading(valuePath),
        original: DbStore.getEditOriginal(id, get(props, "form"), get(props, "prefix", "")),
        user: UserStore.getUser()
      }
    }
  }
}

export default SharedFormEdit
export {
  submitEdit
}
