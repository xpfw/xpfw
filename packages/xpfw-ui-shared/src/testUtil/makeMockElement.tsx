import { cloneDeep, get, isArray, isNil, isObject, isString, set } from "lodash"
import * as React from "react"

const idtmp: string[] = []

const makeMockElement = (name: string, stripIdsAt?: string | string[], idPath: string = "_id") => {
  return class extends React.Component<any, any> {
    public render() {
      const toStringify = this.props
      const stripMe = isArray(stripIdsAt) ? stripIdsAt : [stripIdsAt]
      if (isArray(stripIdsAt)) {
        for (const stripper of stripIdsAt) {
          if (isString(stripper)) {
            let toStrip = get(this.props, stripper)
            if (!isArray(toStrip)) {
              toStrip = [toStrip]
            }
            const newIds = []
            for (const obj of toStrip) {
              const id = isString(obj) ? obj : get(obj, idPath)
              let newId = idtmp.indexOf(id)
              if (newId === -1) {
                newId = idtmp.length + 1
                idtmp.push(id)
              }
              const objClone = cloneDeep(obj)
              if (isString(obj)) {
                newIds.push(newId)
              } else {
                set(objClone, idPath, newId)
                newIds.push(objClone)
              }
            }
            if (newIds.length > 0 && !isNil(newIds[0])) {
              set(toStringify, stripper, newIds)
            }
          }
        }
      }
      return (
        <div className={name}>
          {JSON.stringify(toStringify, undefined, 2)}
        </div>
      )
    }
  }
}

export default makeMockElement
