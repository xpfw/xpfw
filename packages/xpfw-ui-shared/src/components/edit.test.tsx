import { tests } from "@xpfw/ui-tests"
import "isomorphic-fetch"
import makeMockElement from "../testUtil/makeMockElement"
import { submitEdit } from "./edit"

tests.edit(makeMockElement("mymock"), submitEdit)
