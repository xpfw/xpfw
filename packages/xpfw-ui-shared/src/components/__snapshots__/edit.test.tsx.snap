// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbStore Edit Test: After create and geteditorig 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 420,
        "myString": "myText",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {
      "0": Object {
        "result": Object {
          "id": 0,
          "myNum": 420,
          "myString": "myText",
        },
      },
    },
    "users": Object {
      "0": Object {
        "result": Object {
          "email": "user",
          "id": 0,
        },
      },
    },
  },
  "updateState": Object {},
}
`;

exports[`DbStore Edit Test: Before submit patch 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 420,
        "myString": "myText",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {
      "0": Object {
        "result": Object {
          "id": 0,
          "myNum": 420,
          "myString": "myText",
        },
      },
    },
    "users": Object {
      "0": Object {
        "result": Object {
          "email": "user",
          "id": 0,
        },
      },
    },
  },
  "updateState": Object {},
}
`;

exports[`DbStore Edit Test: after create 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "loading": false,
  "original": {
    "result": {
      "myString": "myText",
      "myNum": 420,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: after edit getorig 1`] = `
Object {
  "formData": Object {
    "email": "user",
    "myNum": 420,
    "myString": "myText",
    "password": "",
  },
}
`;

exports[`DbStore Edit Test: after purposely set invalid err 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: after submitted patch 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 420,
        "myString": "myText",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {
      "0": Object {
        "result": Object {
          "id": 0,
          "myNum": 240,
          "myString": "changed",
        },
      },
    },
    "users": Object {
      "0": Object {
        "result": Object {
          "email": "user",
          "id": 0,
        },
      },
    },
  },
  "updateState": Object {
    "0formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 240,
        "myString": "changed",
      },
    },
  },
}
`;

exports[`DbStore Edit Test: after successfull save 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: before set by getorig 1`] = `
Object {
  "formData": Object {
    "email": "user",
    "myNum": null,
    "myString": null,
    "password": "",
  },
}
`;

exports[`DbStore Edit Test: hidden in update 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "hide": {
                "create": false,
                "update": true,
                "find": false
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: hidden text 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "hide": {
                "create": true
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "hide": {
          "create": true
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: not hidden in update 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "hide": {
                "create": true,
                "update": false,
                "find": false
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "hide": {
          "create": true,
          "update": false,
          "find": false
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: prefix After create and geteditorig 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 1,
        "myNum": 24,
        "myString": "secondOBJ",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {
      "0": Object {
        "result": Object {
          "id": 0,
          "myNum": 240,
          "myString": "changed",
        },
      },
      "1": Object {
        "result": Object {
          "id": 1,
          "myNum": 24,
          "myString": "secondOBJ",
        },
      },
    },
    "users": Object {
      "0": Object {
        "result": Object {
          "email": "user",
          "id": 0,
        },
      },
    },
  },
  "updateState": Object {
    "0formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 240,
        "myString": "changed",
      },
    },
    "formModel": undefined,
  },
}
`;

exports[`DbStore Edit Test: prefix Before submit patch 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 1,
        "myNum": 24,
        "myString": "secondOBJ",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {
      "0": Object {
        "result": Object {
          "id": 0,
          "myNum": 240,
          "myString": "changed",
        },
      },
      "1": Object {
        "result": Object {
          "id": 1,
          "myNum": 24,
          "myString": "secondOBJ",
        },
      },
    },
    "users": Object {
      "0": Object {
        "result": Object {
          "email": "user",
          "id": 0,
        },
      },
    },
  },
  "updateState": Object {
    "0formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 240,
        "myString": "changed",
      },
    },
    "formModel": undefined,
  },
}
`;

exports[`DbStore Edit Test: prefix after create 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 1,
  "prefix": "editPrefix",
  "loading": false,
  "original": {
    "result": {
      "myString": "secondOBJ",
      "myNum": 24,
      "id": 1
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: prefix after edit getorig 1`] = `
Object {
  "formData": Object {
    "email": "user",
    "myNum": 24,
    "myString": "secondOBJ",
    "password": "",
  },
}
`;

exports[`DbStore Edit Test: prefix after purposely set invalid err 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 1,
  "prefix": "editPrefix",
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "SECOND change",
      "myNum": 6543,
      "id": 1
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "SECOND change",
      "myNum": 6543,
      "id": 1
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: prefix after submitted patch 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 1,
        "myNum": 24,
        "myString": "secondOBJ",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {
      "0": Object {
        "result": Object {
          "id": 0,
          "myNum": 240,
          "myString": "changed",
        },
      },
      "1": Object {
        "result": Object {
          "id": 1,
          "myNum": 6543,
          "myString": "SECOND change",
        },
      },
    },
    "users": Object {
      "0": Object {
        "result": Object {
          "email": "user",
          "id": 0,
        },
      },
    },
  },
  "updateState": Object {
    "0formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 240,
        "myString": "changed",
      },
    },
    "editPrefix.1formModel": Object {
      "result": Object {
        "id": 1,
        "myNum": 6543,
        "myString": "SECOND change",
      },
    },
    "formModel": undefined,
  },
}
`;

exports[`DbStore Edit Test: prefix after successfull save 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 1,
  "prefix": "editPrefix",
  "state": {
    "result": {
      "myString": "SECOND change",
      "myNum": 6543,
      "id": 1
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "SECOND change",
      "myNum": 6543,
      "id": 1
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: prefix before set by getorig 1`] = `
Object {
  "formData": Object {
    "email": "user",
    "myNum": null,
    "myString": null,
    "password": "",
  },
}
`;

exports[`DbStore Edit Test: prefix result of edit fetch 1`] = `
Object {
  "result": Object {
    "id": 1,
    "myNum": 24,
    "myString": "secondOBJ",
  },
}
`;

exports[`DbStore Edit Test: prefix updated record 1`] = `
Object {
  "result": Object {
    "id": 1,
    "myNum": 6543,
    "myString": "SECOND change",
  },
}
`;

exports[`DbStore Edit Test: prefix while loading 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 1,
  "prefix": "editPrefix",
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "SECOND change",
      "myNum": 6543,
      "id": 1
    }
  },
  "loading": true,
  "original": {
    "result": {
      "myString": "SECOND change",
      "myNum": 6543,
      "id": 1
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: resetOnMount 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "hide": {
                "create": false,
                "update": true,
                "find": false
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "resetOnMount": true,
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": false,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;

exports[`DbStore Edit Test: result of edit fetch 1`] = `
Object {
  "result": Object {
    "id": 0,
    "myNum": 420,
    "myString": "myText",
  },
}
`;

exports[`DbStore Edit Test: updated record 1`] = `
Object {
  "result": Object {
    "id": 0,
    "myNum": 240,
    "myString": "changed",
  },
}
`;

exports[`DbStore Edit Test: while loading 1`] = `
<div
  class="mymock"
>
  {
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "id": 0,
  "error": {
    "errors": [
      "a",
      "b"
    ]
  },
  "state": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "loading": true,
  "original": {
    "result": {
      "myString": "changed",
      "myNum": 240,
      "id": 0
    }
  },
  "fields": [
    {
      "mapTo": "myString",
      "type": 2,
      "validate": {
        "required": {
          "type": 2
        }
      }
    },
    {
      "mapTo": "myNum",
      "type": 3,
      "validate": {}
    }
  ]
}
</div>
`;
